docs/DESIGN.md

```diff
@@ -0,0 +1,472 @@
+timeout, 1+ bytes copied    okay         reset timeout, continue another copy
+```
+
+Ensuring writing is shutdown on the dst conn after seeing an EOF from the
+src conn is intended to send all data read from src before EOF to the dst, and
+then inform the dst that no more messages will be coming.  The `net.Conn`
+abstraction doesn't quite suffice, we also need `CloseWrite` to handle this
+scenario.
+
+```
+dst conn protocol   message sent to inform other party of EOF
+-----------------   -----------------------------------------
+TCP                 FIN
+TLS                 alert close_notify
+```
+
+Ref: "3.5. Closing a Connection" in [rfc793](https://www.ietf.org/rfc/rfc793.txt).
+
+Complication: the "bytes copied" indicator used to express the application
+idle timeout needs to consider bytes copied in both directions, from this
+single copy operation and the anti-parallel one, otherwise it may close
+application connections where one direction of the conversation is silent, but
+the other is very busy. Maybe there's a cleaner way to express it.
+
+### TLS
+
+There is a tradeoff between security and supporting backwards compatibility 
+with clients that may only implement known-vulnerable protocols.
+
+In ascending order of security and decreasing compatibility, we have:
+
+1. server is willing to entertain TLS 1.0 connections, and a wide variety of 
+   cipher suites
+2. server accepts TLS 1.2 or TLS 1.3 only, but refuses to use particularly 
+   broken cipher suites
+3. server accepts TLS 1.3 only, limited to the
+   [set of cipher suites](https://datatracker.ietf.org/doc/html/rfc8446#section-9.1)
+   that the TLS 1.3 standard dictates TLS-compliant applications MUST or
+   SHOULD support
+4. custom protocol: server accepts connections with TLS 1.3 only, using a 
+   fixed choice of cipher suites (`TLS_CHACHA20_POLY1305_SHA256`) and curve  
+   preferences (`X25519`), and will refuse to accept anything less, as 
+   recommended by [a latacora blog post from 2018](https://latacora.micro.blog/2018/04/03/cryptographic-right-answers.html).
+
+For the proof-of-concept, the server & a compatible client will restrict 
+themselves to option 3, TLS 1.3, as that is simple to configure, and may 
+implement option 4 if time permits.
+
+### Authentication
+
+Authentication will be implemented using TLS with mutual authentication 
+between the client and the server.
+
+The server must be configured with a certificate and corresponding private
+key. This certificate will be presented to clients who wish to negotiate TLS
+connections.
+
+The server will load trusted CA certs from the environment in the usual 
+operating-system defined location. These CA certs will be used to validate 
+the trust chain of the certificate presented by the client. Since the client 
+certificate will be used as the basis of authentication, the server MUST NOT 
+be configured to trust CA roots that are not trusted to verify client 
+identities for purposes of authentication.
```

Comment:
Could you elaborate on how users will configure which CAs to trust vs not trust for client authentication?

Reply:
excellent point, can do

Reply:
done below

Reply:
Apologies, I meant specifically on the server side. If trusted CAs are loaded from "the usual operating-system defined location" and if only a subset of those certificates are trusted to authenticate clients, how will the server know which CAs to trust and which not to?

Reply:
ah, i understand. your concern makes complete sense.  in containerised deployments etc a user could configure the container to only contain certificates that should be trusted by the proxy as CAs for purposes of verifying clients. but even that is a bit awkward if the server or other processes in the container (if any) need to integrate with and authenticate other external systems

i will rework the CLI UX to force the user of the server to specify a custom location of the CA roots used to validate the client's certificate.  that adds a little friction but makes the decision explicit and is safer.

Reply:
Thanks for the clarification, that makes sense!

Reply:
created https://github.com/fcostin/tcplb/issues/5 to track, will follow up with a PR for next week